File: cli/src/freepod/project.py
"""The `.freepod.json` project file: load, save, and project-root discovery.

The file holds intent — the environment, the deployment pointer, and the user
values — and nothing a deploy would rewrite. In particular the build's `image`
is never written to it: it is a build output, not intent, and persisting it
would mean a rewritten committed file on every deploy, which is git churn and a
merge conflict for any team of two. See design D4.

```json
{
  "version": 1,
  "env": "prod",
  "deployment": {"id": "40bd8dea-…", "name": "custom-d8dtx4"},
  "user_values": {"hostname": "myapp.freepod.eu"}
}
```

Before the first deploy, `deployment` is `null`.
"""

from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Any, Dict, Optional

from . import FreepodError, UsageError

PROJECT_FILE = ".freepod.json"

#: Bumped when the on-disk format changes incompatibly.
FORMAT_VERSION = 1

#: Keys a deploy produces rather than a user declaring. Stripped on write.
#:
#: `image` is not merely omitted when absent — it is removed even if a caller
#: passes one, because the platform's schema declares it `type: "string"` under
#: `additionalProperties: false`, so neither a value nor an explicit null
#: belongs in a file that is committed and diffed.
BUILD_OUTPUT_KEYS = frozenset({"image"})


class Project:
    """One project's file, and the directory it was found in."""

    def __init__(
        self,
        root: Path,
        env: str,
        user_values: Optional[Dict[str, Any]] = None,
        deployment: Optional[Dict[str, str]] = None,
        version: int = FORMAT_VERSION,
    ):
        self.root = Path(root)
        self.env = env
        self.user_values = dict(user_values or {})
        self.deployment = dict(deployment) if deployment else None
        self.version = version

    # -- location ---------------------------------------------------------

    @property
    def path(self) -> Path:
        return self.root / PROJECT_FILE

    @property
    def deployment_id(self) -> Optional[str]:
        return self.deployment.get("id") if self.deployment else None

    @property
    def deployment_name(self) -> Optional[str]:
        return self.deployment.get("name") if self.deployment else None

    @property
    def hostname(self) -> Optional[str]:
        value = self.user_values.get("hostname")
        return value if isinstance(value, str) else None

    # -- persistence ------------------------------------------------------

    def to_document(self) -> Dict[str, Any]:
        values = {k: v for k, v in self.user_values.items() if k not in BUILD_OUTPUT_KEYS}
        return {
            "version": self.version,
            "env": self.env,
            "deployment": dict(self.deployment) if self.deployment else None,
            "user_values": values,
        }

    def save(self) -> None:
        """Write the file, atomically, with a trailing newline.

        Atomic-replace so an interrupted write cannot truncate a file that
        carries the only pointer to a live deployment.
        """
        document = self.to_document()
        temporary = self.path.with_name(f"{PROJECT_FILE}.{os.getpid()}.tmp")
        try:
            with open(temporary, "w", encoding="utf-8") as handle:
                json.dump(document, handle, indent=2)
                handle.write("\n")
            os.replace(temporary, self.path)
        except BaseException:
            try:
                os.unlink(temporary)
            except OSError:
                pass
            raise

    def record_deployment(self, deployment_id: str, name: str) -> None:
        """Pin the deployment pointer and persist it immediately.

        Written before the rollout is awaited: a deployment that exists but is
        not recorded is one the user cannot see, address, or delete.
        """
        self.deployment = {"id": str(deployment_id), "name": name}
        self.save()

    def forget_deployment(self) -> None:
        """Drop the deployment pointer and persist it immediately.

        The counterpart of `record_deployment`, and written the moment the
        platform accepts a deletion rather than once the teardown lands: from
        that point the deployment can never serve this project again, so a
        pointer to it would only make the next deploy fail. Everything else in
        the file — the environment, the user values — is intent and survives,
        so a later `freepod deploy` re-creates under the same hostname.
        """
        self.deployment = None
        self.save()


# --------------------------------------------------------------------------
# Discovery
# --------------------------------------------------------------------------


def find_project_root(start: Optional[Path] = None) -> Optional[Path]:
    """The nearest ancestor of `start` containing `.freepod.json`, git-style."""
    current = Path(start or Path.cwd()).resolve()
    for directory in [current, *current.parents]:
        if (directory / PROJECT_FILE).is_file():
            return directory
    return None


def load(root: Path) -> Project:
    """Read the project file in `root`."""
    path = Path(root) / PROJECT_FILE
    try:
        with open(path, "r", encoding="utf-8") as handle:
            document = json.load(handle)
    except OSError as exc:
        raise FreepodError(f"cannot read {path}: {exc}") from None
    except ValueError as exc:
        raise FreepodError(f"{path} is not valid JSON: {exc}") from None

    if not isinstance(document, dict):
        raise FreepodError(f"{path} must contain a JSON object")

    version = document.get("version", FORMAT_VERSION)
    if not isinstance(version, int) or version > FORMAT_VERSION:
        raise FreepodError(
            f"{path} declares format version {version!r}, which this client does not "
            f"understand (it supports up to {FORMAT_VERSION}) — upgrade freepod"
        )

    env = document.get("env")
    if not isinstance(env, str) or not env:
        raise FreepodError(f"{path} does not record which environment it belongs to")

    user_values = document.get("user_values")
    if user_values is None:
        user_values = {}
    if not isinstance(user_values, dict):
        raise FreepodError(f"{path}: 'user_values' must be an object")

    deployment = document.get("deployment")
    if deployment is not None:
        if not isinstance(deployment, dict) or not deployment.get("id"):
            raise FreepodError(f"{path}: 'deployment' must be null or carry an 'id'")

    return Project(
        root=Path(root),
        env=env,
        user_values=user_values,
        deployment=deployment,
        version=version,
    )


def require_project(env_name: str, start: Optional[Path] = None) -> Project:
    """Find and load the project, refusing an environment mismatch.

    The environment check is the same audience-scoping problem as the token
    cache, one level up: a deployment id minted on dev is meaningless on prod,
    so a mismatch is an error rather than something to guess at.
    """
    root = find_project_root(start)
    if root is None:
        where = Path(start or Path.cwd()).resolve()
        raise UsageError(
            f"no {PROJECT_FILE} found in {where} or any parent directory — "
            f"this project is not initialized. Run `freepod init` to set it up."
        )

    project = load(root)
    if project.env != env_name:
        raise UsageError(
            f"{project.path} belongs to the '{project.env}' environment, but this "
            f"command targets '{env_name}'.\n"
            f"  Re-run with --env {project.env}, or work in a project initialized "
            f"for '{env_name}'."
        )
    return project

File: cli-rust/src/project.rs
//! The `.freepod-rust.json` project file: load, save, and project-root discovery.
//!
//! Mirrors `project.py`. The file holds intent — the environment, the
//! deployment pointer, and the user values — and nothing a deploy would
//! rewrite. The Rust client uses a distinct filename so the two clients can
//! coexist in one project directory.

use std::collections::HashMap;
use std::path::{Path, PathBuf};

use serde_json::{json, Value};

use crate::errors::{freepod, usage, Result};

pub const PROJECT_FILE: &str = ".freepod-rust.json";

/// Bumped when the on-disk format changes incompatibly.
pub const FORMAT_VERSION: u64 = 1;

/// Keys a deploy produces rather than a user declaring. Stripped on write.
const BUILD_OUTPUT_KEYS: &[&str] = &["image"];

/// One project's file, and the directory it was found in.
#[derive(Debug, Clone)]
pub struct Project {
    pub root: PathBuf,
    pub env: String,
    pub user_values: HashMap<String, Value>,
    pub deployment: Option<HashMap<String, String>>,
    pub version: u64,
}

impl Project {
    pub fn new(
        root: PathBuf,
        env: impl Into<String>,
        user_values: HashMap<String, Value>,
    ) -> Self {
        Self {
            root,
            env: env.into(),
            user_values,
            deployment: None,
            version: FORMAT_VERSION,
        }
    }

    pub fn path(&self) -> PathBuf {
        self.root.join(PROJECT_FILE)
    }

    pub fn deployment_id(&self) -> Option<&str> {
        self.deployment.as_ref().and_then(|d| d.get("id").map(|s| s.as_str()))
    }
    pub fn deployment_name(&self) -> Option<&str> {
        self.deployment.as_ref().and_then(|d| d.get("name").map(|s| s.as_str()))
    }

    /// The document as written to disk.
    pub fn to_document(&self) -> Value {
        let values: HashMap<String, Value> = self
            .user_values
            .iter()
            .filter(|(k, _)| !BUILD_OUTPUT_KEYS.contains(&k.as_str()))
            .map(|(k, v)| (k.clone(), v.clone()))
            .collect();
        json!({
            "version": self.version,
            "env": self.env,
            "deployment": self.deployment.as_ref().map(|d| {
                let mut obj = serde_json::Map::new();
                for (k, v) in d {
                    obj.insert(k.clone(), Value::String(v.clone()));
                }
                Value::Object(obj)
            }),
            "user_values": values,
        })
    }

    /// Write the file, atomically, with a trailing newline.
    pub fn save(&self) -> Result<()> {
        let document = self.to_document();
        let text = serde_json::to_string_pretty(&document)
            .map_err(|e| freepod(format!("cannot serialize project file: {e}")))?;
        let text = format!("{text}\n");
        let path = self.path();
        let pid = std::process::id();
        let temporary = path.with_file_name(format!("{PROJECT_FILE}.{pid}.tmp"));
        match std::fs::write(&temporary, text) {
            Ok(()) => std::fs::rename(&temporary, &path)
                .map_err(|e| {
                    let _ = std::fs::remove_file(&temporary);
                    freepod(format!("cannot write {}: {e}", path.display()))
                }),
            Err(e) => {
                let _ = std::fs::remove_file(&temporary);
                Err(freepod(format!("cannot write {}: {e}", path.display())))
            }
        }
    }

    /// Pin the deployment pointer and persist it immediately.
    pub fn record_deployment(&mut self, deployment_id: impl Into<String>, name: impl Into<String>) -> Result<()> {
        let mut d = HashMap::new();
        d.insert("id".to_string(), deployment_id.into());
        d.insert("name".to_string(), name.into());
        self.deployment = Some(d);
        self.save()
    }

    /// Drop the deployment pointer and persist it immediately.
    pub fn forget_deployment(&mut self) -> Result<()> {
        self.deployment = None;
        self.save()
    }
}

/// The nearest ancestor of `start` containing the project file, git-style.
pub fn find_project_root(start: Option<&Path>) -> Option<PathBuf> {
    let current = match start {
        Some(p) => p.to_path_buf(),
        None => std::env::current_dir().ok()?,
    };
    let current = current.canonicalize().ok()?;
    let mut dir: Option<PathBuf> = Some(current);
    while let Some(d) = dir {
        if d.join(PROJECT_FILE).is_file() {
            return Some(d);
        }
        dir = d.parent().map(|p| p.to_path_buf());
    }
    None
}

/// Read the project file in `root`.
pub fn load(root: &Path) -> Result<Project> {
    let path = root.join(PROJECT_FILE);
    let raw = std::fs::read_to_string(&path)
        .map_err(|e| freepod(format!("cannot read {}: {e}", path.display())))?;
    let document: Value = serde_json::from_str(&raw)
        .map_err(|e| freepod(format!("{path:?} is not valid JSON: {e}")))?;
    let document = match document {
        Value::Object(_) => document,
        _ => return Err(freepod(format!("{path:?} must contain a JSON object"))),
    };

    let version = match document.get("version") {
        None => FORMAT_VERSION,
        Some(Value::Number(n)) => {
            let v = n.as_u64().unwrap_or(0);
            if v > FORMAT_VERSION {
                return Err(freepod(format!(
                    "{:?} declares format version {v}, which this client does not understand \
                     (it supports up to {FORMAT_VERSION}) — upgrade freepod",
                    path
                )));
            }
            v
        }
        Some(other) => {
            return Err(freepod(format!(
                "{:?} declares format version {other}, which this client does not understand \
                 (it supports up to {FORMAT_VERSION}) — upgrade freepod",
                path
            )))
        }
    };

    let env = match document.get("env") {
        Some(Value::String(s)) if !s.is_empty() => s.clone(),
        _ => {
            return Err(freepod(format!(
                "{:?} does not record which environment it belongs to",
                path
            )))
        }
    };

    let user_values = match document.get("user_values") {
        None => HashMap::new(),
        Some(Value::Object(obj)) => obj
            .iter()
            .map(|(k, v)| (k.clone(), v.clone()))
            .collect(),
        Some(_) => {
            return Err(freepod(format!("{:?}: 'user_values' must be an object", path)))
        }
    };

    let deployment = match document.get("deployment") {
        None => None,
        Some(Value::Null) => None,
        Some(Value::Object(obj)) => {
            if obj.get("id").and_then(|v| v.as_str()).map(|s| !s.is_empty()).unwrap_or(false) {
                Some(
                    obj.iter()
                        .filter_map(|(k, v)| v.as_str().map(|s| (k.clone(), s.to_string())))
                        .collect(),
                )
            } else {
                return Err(freepod(format!(
                    "{:?}: 'deployment' must be null or carry an 'id'",
                    path
                )));
            }
        }
        Some(_) => {
            return Err(freepod(format!(
                "{:?}: 'deployment' must be null or carry an 'id'",
                path
            )))
        }
    };

    Ok(Project {
        root: root.to_path_buf(),
        env,
        user_values,
        deployment,
        version,
    })
}

/// Find and load the project, refusing a directory that has none.
pub fn require_project(start: Option<&Path>) -> Result<Project> {
    let root = match find_project_root(start) {
        Some(r) => r,
        None => {
            let where_ = match start {
                Some(p) => p.to_path_buf(),
                None => std::env::current_dir().unwrap_or_default(),
            }
            .canonicalize()
            .unwrap_or_else(|_| std::env::current_dir().unwrap_or_default());
            return Err(usage(format!(
                "no {PROJECT_FILE} found in {} or any parent directory — \
                 this project is not initialized. Run `freepod init` to set it up.",
                where_.display()
            )));
        }
    };
    load(&root)
}
